File: 0079/recursive.cpp
#include <cassert>
#include <cstdlib>
#include <string>
#include <vector>

#include "gtest/gtest.h"

using namespace std;

class Solution {
 public:
  bool exist(vector<vector<char>>& board, string word) {
    for (int row_idx = 0; row_idx < board.size(); row_idx++) {
      for (int col_idx = 0; col_idx < board[row_idx].size(); col_idx++) {
        if (board[row_idx][col_idx] != word[0]) {
          continue;
        }
        if (recursive(board, row_idx, col_idx, word, 0)) {
          return true;
        }
      }
    }
    return false;
  }

 private:
  bool recursive(vector<vector<char>>& board, int row_idx, int col_idx, const string& word, size_t word_idx) {
    if (row_idx < 0 || row_idx >= board.size() || col_idx < 0 || col_idx >= board[row_idx].size()) {
      return false;
    }
    if (board[row_idx][col_idx] != word[word_idx]) {
      return false;
    }
    if (word_idx == word.size() - 1) {
      return true;
    }
    const char temp = board[row_idx][col_idx];
    board[row_idx][col_idx] = '#';
    bool found = recursive(board, row_idx+1, col_idx, word, word_idx+1) ||
                 recursive(board, row_idx-1, col_idx, word, word_idx+1) ||
                 recursive(board, row_idx, col_idx+1, word, word_idx+1) ||
                 recursive(board, row_idx, col_idx-1, word, word_idx+1);
    board[row_idx][col_idx] = temp;
    return found;
  }
};

int main() {
  Solution sol;
  vector<vector<char>> board;
  string word;

  board = {
    {'A','B','C','E'},
    {'S','F','C','S'},
    {'A','D','E','E'}
  };
  word = "ABCCED";
  EXPECT_TRUE(sol.exist(board, word));
  word = "SEE";
  EXPECT_TRUE(sol.exist(board, word));
  word = "ABCB";
  EXPECT_FALSE(sol.exist(board, word));

  exit(EXIT_SUCCESS);
}

File: 0079/recursive.rs
struct Solution;

impl Solution {
    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        let (m, n) = (board.len(), board[0].len());
        let word_as_bytes = word.as_bytes();
        let mut visited = vec![vec![false; n]; m];
        for i in 0..m {
            for j in 0..n {
                if Solution::recursive(&board, i, j, word_as_bytes, 0, &mut visited) {
                    return true;
                }
            }
        }
        return false;
    }

    fn recursive(board: &Vec<Vec<char>>, i: usize, j: usize, word_as_bytes: &[u8], word_idx: usize, visited: &mut Vec<Vec<bool>>) -> bool {
        if visited[i][j] || board[i][j] != word_as_bytes[word_idx] as char {
            return false;
        }
        if word_idx == word_as_bytes.len() - 1 {
            return true;
        }
        visited[i][j] = true;
        let mut ret = false;
        let (m, n) = (board.len(), board[0].len());
        if i < m - 1 {
            ret = ret || Solution::recursive(&board, i + 1, j, word_as_bytes, word_idx + 1, visited);
        }
        if i > 0 {
            ret = ret || Solution::recursive(&board, i - 1, j, word_as_bytes, word_idx + 1, visited);
        }
        if j < n - 1 {
            ret = ret || Solution::recursive(&board, i, j + 1, word_as_bytes, word_idx + 1, visited);
        }
        if j > 0 {
            ret = ret || Solution::recursive(&board, i, j - 1, word_as_bytes, word_idx + 1, visited);
        }
        visited[i][j] = false;
        return ret;
    }
}

fn main() {
    let board = vec![vec!['A', 'B', 'C', 'E'], vec!['S', 'F', 'C', 'S'], vec!['A', 'D', 'E', 'E']];
    let word = String::from("ABCCED");
    let expected = true;
    assert_eq!(Solution::exist(board, word), expected);

    let board = vec![vec!['A', 'B', 'C', 'E'], vec!['S', 'F', 'C', 'S'], vec!['A', 'D', 'E', 'E']];
    let word = String::from("SEE");
    let expected = true;
    assert_eq!(Solution::exist(board, word), expected);

    let board = vec![vec!['A', 'B', 'C', 'E'], vec!['S', 'F', 'C', 'S'], vec!['A', 'D', 'E', 'E']];
    let word = String::from("ABCB");
    let expected = false;
    assert_eq!(Solution::exist(board, word), expected);
}
